File: integration-tests/src/test/java/dev/capylang/test/LocalFunctionsTest.java
package dev.capylang.test;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class LocalFunctionsTest {
    @Test
    void localRecursiveFunctionCanBeCalled() {
        assertThat(LocalFunctions.sumDown(0)).isEqualTo(0);
        assertThat(LocalFunctions.sumDown(1)).isEqualTo(1);
        assertThat(LocalFunctions.sumDown(5)).isEqualTo(15);
    }

    @Test
    void localMutualRecursionWorks() {
        assertThat(LocalFunctions.parity(0)).isTrue();
        assertThat(LocalFunctions.parity(1)).isFalse();
        assertThat(LocalFunctions.parity(12)).isTrue();
        assertThat(LocalFunctions.parity(15)).isFalse();
    }

    @Test
    void localFunctionCanUseLetExpression() {
        assertThat(LocalFunctions.localWithLet(0)).isEqualTo(1);
        assertThat(LocalFunctions.localWithLet(10)).isEqualTo(11);
    }
}

File: capy/src/e2e-cfun/js/dev/capylang/test/LocalFunctions.test.js
const test = require('node:test');
const { assert, generatedModule } = require('./_support');

test('LocalFunctions', () => {
    const localFunctions = generatedModule('dev/capylang/test/LocalFunctions.js');
    assert.equal(localFunctions.sumDown(5), 15);
    assert.equal(localFunctions.parity(15), false);
    assert.equal(localFunctions.localWithLet(10), 11);
    assert.equal(localFunctions.maxLocal(2, 5), 5);
});
